numpy/linalg/linalg.py

```diff
@@ -596,47 +585,56 @@ def matrix_power(M, n):
            [ 1.,  0.,  0.,  0.],
            [ 0.,  0.,  0.,  1.],
            [ 0.,  0., -1.,  0.]])
-    >>> LA.matrix_power(q, 2) # = -np.eye(4)
+    >>> matrix_power(q, 2) # = -np.eye(4)
     array([[-1.,  0.,  0.,  0.],
            [ 0., -1.,  0.,  0.],
            [ 0.,  0., -1.,  0.],
            [ 0.,  0.,  0., -1.]])
 
     """
-    M = asanyarray(M)
-    if M.ndim != 2 or M.shape[0] != M.shape[1]:
+    a = array(a, subok=True, copy=False, ndmin=1)
+    m = a.shape[-1]
+    if a.ndim < 2 or a.shape[-2] != m:
         raise ValueError("input must be a square array")
-    if not issubdtype(type(n), integer):
+
+    try:
+        n = operator.index(n)
+    except TypeError:
         raise TypeError("exponent must be an integer")
 
-    from numpy.linalg import inv
+    if n == 0:
+        a = empty_like(a)
+        a[...] = eye(m, dtype=a.dtype)
+        return a
 
-    if n==0:
-        M = M.copy()
-        M[:] = identity(M.shape[0])
-        return M
-    elif n<0:
-        M = inv(M)
-        n *= -1
+    elif n < 0:
+        a = inv(a)
+        n = abs(n)
 
-    result = M
-    if n <= 3:
-        for _ in range(n-1):
-            result=dot(result, M)
-        return result
+    # short-cuts.
+    if n == 1:
+        return a
+
+    elif n == 2:
+        return matmul(a, a)
+
+    elif n == 3:
+        return matmul(matmul(a, a), a)
+
+    # binary decomposition to reduce the number of Matrix multiplications.
+    z = a
+    left, use = divmod(n, 2)
+    while not use:
+        z = matmul(z, z)
+        left, use = divmod(left, 2)
+
+    result = z
+    while left > 0:
+        z = matmul(z, z)
+        left, use = divmod(left, 2)
+        if use:
+            result = matmul(result, z)
```

Comment:
Way better without going through a string!

Could consider also writing:
```python
z = a
while True:
    n, use = divmod(n, 2)
    if use:
        if result is None:
            result = z  # optimization - don't initialize result to the identity
        else:
            result = matmul(result, z)
    if not n:
        break

    z = matmul(z, z)
```

which makes it a bit more obvious that two loops are only needed to avoid a multiplication with the identity